docs/howto/migrate/migration-guide.md

```diff
@@ -0,0 +1,1163 @@
+          } yield ()
+      }
+  }
+```
+
+ZIO 2.x deprecates all `ZLayer.fromService*` functions:
+
+| ZIO 1.0                          | ZIO 2.x   |
+|----------------------------------|-----------|
+| `ZLayer.fromService`             | `toLayer` |
+| `ZLayer.fromServices`            | `toLayer` |
+| `ZLayer.fromServiceM`            | `toLayer` |
+| `ZLayer.fromServicesM`           | `toLayer` |
+| `ZLayer.fromServiceManaged`      | `toLayer` |
+| `ZLayer.fromServicesManaged`     | `toLayer` |
+| `ZLayer.fromServiceMany`         | `toLayer` |
+| `ZLayer.fromServicesMany`        | `toLayer` |
+| `ZLayer.fromServiceManyM`        | `toLayer` |
+| `ZLayer.fromServicesManyM`       | `toLayer` |
+| `ZLayer.fromServiceManyManaged`  | `toLayer` |
+| `ZLayer.fromServicesManyManaged` | `toLayer` |
+
+Instead, it provides the `toLayer` extension methods for functions:
+
+```scala
+case class LoggingLive(console: Console, clock: Clock) extends Logging {
+  override def log(line: String): UIO[Unit] =
+    for {
+      current <- clock.currentDateTime.orDie
+      _       <- console.putStrLn(current.toString + "--" + line).orDie
+    } yield ()
+}
+
+object LoggingLive {
+  val layer: URLayer[Has[Console] with Has[Clock], Has[Logging]] =
+    (LoggingLive(_, _)).toLayer[Logging]
+}
+```
+
+Note that the `LoggingLive(_, _)` is a `Function2` of type `(Console, Clock) => LoggingLive`. As the ZIO 2.x provides the `toLayer` extension method for all `Function` arities, we can call the `toLayer` on any function to convert that to the `ZLayer`. Unlike the `ZLayer.fromService*` functions, this can completely infer the input types, so it saves us from a lot of boilerplates we have had in ZIO 1.x.
+
+### Accessing a Service from the Environment
+
+Assume we have a service named `Logging`:
+
+```scala mdoc:silent
+trait Logging {
+  def log(line: String): UIO[Unit]
+}
+```
+
+In ZIO 1.x, when we wanted to access a service from the environment, we used the `ZIO.access` + `Has#get` combination (`ZIO.access(_.get)`):
+
+```scala mdoc:silent:nest
+val logging: URIO[Has[Logging], Logging] = ZIO.access(_.get)
+```
+
+Also, to create accessor methods, we used the following code:
+
+```scala mdoc:silent:nest
+def log(line: String): URIO[Has[Logging], Unit] = ZIO.accessM(_.get.log(line))
+```
+
+ZIO 2.x reduces one level of indirection by using `ZIO.service` operator:
+
+```scala mdoc:silent:nest
+val logging : URIO[Has[Logging], Logging] = ZIO.service
+```
+
+And to write the accessor method in ZIO 2.x, we can use `ZIO.serviceWith` operator:
+
+```scala mdoc:silent:nest
+def log(line: String): URIO[Has[Logging], Unit] = ZIO.serviceWith(_.log(line))
+```
+### Accessing Multiple Services in the Environment
+
+In ZIO 1.x, we could access multiple services using higher arity service accessors like `ZIO.services` and `ZManaged.services`:
+
+```scala mdoc:silent:nest
+for {
+  (console, random) <- ZIO.services[Console, Random]
```

Comment:
hmm.. I think this isn't working as expected cause of the `withFilter` thingy.. e.g. it was always 

```scala
for { 
  services <- ZIO.services[Console, Random]
  (console, random) = services
  ...
  }
  ```
  
  not sure if it matters in the migration guide, though.

Reply:
> hmm.. I think this isn't working as expected cause of the withFilter thingy.. e.g. it was always

Does the `withFilter` method of for-comprehension cause any problem in this case?

> not sure if it matters in the migration guide, though.

The `ZIO.services` is deprecated in ZIO 2.x so we should cover that in the migration guide.